lib/context-manager/legacy-context-manager.js

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2021 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+/**
+ * Class for managing state in the agent.
+ * Keeps track of a single context instance.
+ *
+ * Given current usage with every instrumented function, the functions in this
+ * class should do as little work as possible to avoid unnecessary overhead.
+ *
+ * @class
+ */
+class LegacyContextManager {
+  /**
+   * @param {object} config New Relic config instance
+   */
+  constructor(config) {
+    this._config = config
+    this._context = null
+  }
+
+  /**
+   * Get the currently active context.
+   *
+   * @returns {object} The current active context.
+   */
+  getContext() {
+    return this._context
+  }
+
+  /**
+   * Set a new active context. Not bound to function execution.
+   *
+   * @param {object} newContext The context to set as active.
+   */
+  setContext(newContext) {
+    this._context = newContext
+  }
+
+  /**
+   * Run a function with the passed in context as the active context.
+   * Restores the previously active context upon completion.
+   *
+   * @param {object} context The context to set as active during callback execution.
+   * @param {Function} callback The function to execute in context.
+   * @param {Function} [cbThis] Optional `this` to apply to the callback.
+   * @param {Array<*>} [args] Optional arguments object or args array to invoke the callback with.
+   * @returns {*} Returns the value returned by the callback function.
+   */
+  runInContext(context, callback, cbThis, args) {
+    const oldContext = this._context
+    this.setContext(context)
+
+    try {
+      return callback.apply(cbThis, args)
```

Comment:
What happens if the callback is async? Could another function swap out context underneath it?

Reply:
Yes, which is actually how the current state of things is designed to be used. You'll see we bind context to callbacks, etc. to restore proper segment parenting as things jump around. So at the end of synchronous execution (a normal method invocation) we do want to restore the parent so anything else jumping in via event loop is starting from the appropriate place, etc.

As we migrate to the newer mechanisms, this will all be stored on the current async resource which makes it a little less funky when reading/pondering. It will also mean we are doing too much work in several places.

For reference, I'd take a close look at the places I've swapped in the `runInContext` and see how they are written, potentially look at how bindSegment() wraps callbacks/promise stuff and maybe the timers. It is not easy to follow at all. Happy to deep-dive as summarizing the current setup cuts out a lot.

Reply:
Great, thanks for the clarification. It is tough to reason about but I see what you're saying. Thanks for the pointers :)